### ISSUE TITLE
TLS communication breaks when deployed an nginx-based application behind istio

### ISSUE THREAD

**Issue by pantelis-karamolegkos** [istio/istio#23422]
**Bug description**

Exposing a service (whose container uses `nginx`) behind envoy, causes cryptic TLS error and connection is terminated.

**Expected behavior**

I have an application that uses `istio`.

Its virtual service is configured to listen to `myservice.myurl.com`.

```
▶ k get vs
NAME                     GATEWAYS                           HOSTS                                                     AGE
my-virtual-service   [myns/istio-ingress-gateway]   [myservice.myurl.com]                7d
```
 
Pod and its sidecar are healthy

```
▶ k get pods | grep myapp
myapp-758bc87dcf-tp5nv                        2/2     Running   0          22m
```

The app is exposed on port `8080` on the container, using `nginx` with the following config

```
worker_processes 4;
pid /tmp/nginx.pid;

daemon off;
error_log /dev/stderr info;

events {
        worker_connections 1024;
}

http {
        client_body_temp_path /tmp/client_body;
        fastcgi_temp_path /tmp/fastcgi_temp;
        proxy_temp_path /tmp/proxy_temp;
        scgi_temp_path /tmp/scgi_temp;
        uwsgi_temp_path /tmp/uwsgi_temp;

        include /etc/nginx/mime.types;

        gzip on;
        gzip_disable "msie6";

        log_format main 'ip=\$http_x_real_ip [\$time_local] '
        '"\$request" \$status \$body_bytes_sent "\$http_referer" '
        '"\$http_user_agent"' ;

        log_format scripts '$document_root | $uri | > $request';

        default_type application/octet-stream;

        server {
                listen 8080 default_server;


                access_log /dev/stdout main;
                access_log /dev/stdout scripts;

                location ~ /api {
                        include /etc/nginx/uwsgi_params;
                        uwsgi_pass unix:/tmp/uwsgi.sock;

                        proxy_set_header Host $host:$server_port;
                        proxy_set_header X-Real-IP $remote_addr;
                        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                }

                root /web;
                index index.html;
                location / {
                        try_files $uri $uri/ /index.html;
                }
        }
}
```

When I try to access the url that my virtual service is pointing to

I get the following error

```
▶ curl -kv https://myservice.myurl.com
*   Trying 35.245.64.86...
* TCP_NODELAY set
* Connected to myservice.myurl.com (35.245.64.86) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-CHACHA20-POLY1305
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=*.myurl.com
*  start date: Apr 10 06:09:43 2020 GMT
*  expire date: Jul  9 06:09:43 2020 GMT
*  issuer: CN=Fake LE Intermediate X1
*  SSL certificate verify result: unable to get local issuer certificate (20), continuing anyway.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x7fd970003a00)
> GET / HTTP/2
> Host: myservice.myurl.com
> User-Agent: curl/7.64.1
> Accept: */*
>
* Connection state changed (MAX_CONCURRENT_STREAMS == 4294967295)!
< HTTP/2 503
< content-length: 95
< content-type: text/plain
< date: Thu, 30 Apr 2020 15:34:52 GMT
< server: istio-envoy
<
* Connection #0 to host myservice.myurl.com left intact
upstream connect error or disconnect/reset before headers. reset reason: connection termination* Closing connection 0
```

The error seems to come from the `istio` sidecar.

When I exec into a (non-istio enabled) pod and `curl` the service directly:

```
/ # curl -kv http://myapp
* Rebuilt URL to: http://myapp/
*   Trying 10.8.58.228...
* TCP_NODELAY set
* Connected to myapp (10.8.58.228) port 80 (#0)
> GET / HTTP/1.1
> Host: myapp
> User-Agent: curl/7.57.0
> Accept: */*
>
* Recv failure: Connection reset by peer
* Closing connection 0
curl: (56) Recv failure: Connection reset by peer
```

Execing into the `envoy` proxy of the problematic pod and turning trace level logging on, I see this error

```
serviceB-758bc87dcf-jzjgj istio-proxy [2020-04-24 13:15:21.180][29][debug][connection] [external/envoy/source/extensions/transport_sockets/tls/ssl_socket.cc:168] [C1484] handshake error: 1
serviceB-758bc87dcf-jzjgj istio-proxy [2020-04-24 13:15:21.180][29][debug][connection] [external/envoy/source/extensions/transport_sockets/tls/ssl_socket.cc:201] [C1484] TLS error: 268435612:SSL routines:OPENSSL_internal:HTTP_REQUEST
```

Nginx logs of the app, are intact.

What could be causing this?

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**

```
▶ istioctl version --remote
client version: version.BuildInfo{Version:"1.1.0", GitRevision:"82797c0c0649a3f73029b33957ae105260458c6e", User:"root", Host:"996cd064-49c1-11e9-813c-0a580a2c0506", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.0-rc.6"}
citadel version: version.BuildInfo{Version:"1.2.0", GitRevision:"f8295503296ec6bae1b1047cc1491469d5e72754-dirty", User:"root", Host:"e1b1b59e-920e-11e9-9590-0a580a2c0406", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.2.0-rc.2-12-gf829550"}
galley version: version.BuildInfo{Version:"1.2.0", GitRevision:"f8295503296ec6bae1b1047cc1491469d5e72754-dirty", User:"root", Host:"e1b1b59e-920e-11e9-9590-0a580a2c0406", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.2.0-rc.2-12-gf829550"}
ingressgateway version: version.BuildInfo{Version:"1.2.0", GitRevision:"f8295503296ec6bae1b1047cc1491469d5e72754", User:"root", Host:"e1b1b59e-920e-11e9-9590-0a580a2c0406", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.2.0-rc.2-12-gf829550"}
ingressgateway version: version.BuildInfo{Version:"1.2.0", GitRevision:"f8295503296ec6bae1b1047cc1491469d5e72754", User:"root", Host:"e1b1b59e-920e-11e9-9590-0a580a2c0406", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.2.0-rc.2-12-gf829550"}
pilot version: version.BuildInfo{Version:"1.2.0", GitRevision:"f8295503296ec6bae1b1047cc1491469d5e72754-dirty", User:"root", Host:"e1b1b59e-920e-11e9-9590-0a580a2c0406", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.2.0-rc.2-12-gf829550"}
pilot version: version.BuildInfo{Version:"1.2.0", GitRevision:"f8295503296ec6bae1b1047cc1491469d5e72754-dirty", User:"root", Host:"e1b1b59e-920e-11e9-9590-0a580a2c0406", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.2.0-rc.2-12-gf829550"}
policy version: version.BuildInfo{Version:"1.2.0", GitRevision:"f8295503296ec6bae1b1047cc1491469d5e72754-dirty", User:"root", Host:"e1b1b59e-920e-11e9-9590-0a580a2c0406", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.2.0-rc.2-12-gf829550"}
sidecar-injector version: version.BuildInfo{Version:"1.2.0", GitRevision:"f8295503296ec6bae1b1047cc1491469d5e72754-dirty", User:"root", Host:"e1b1b59e-920e-11e9-9590-0a580a2c0406", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.2.0-rc.2-12-gf829550"}
telemetry version: version.BuildInfo{Version:"1.2.0", GitRevision:"f8295503296ec6bae1b1047cc1491469d5e72754-dirty", User:"root", Host:"e1b1b59e-920e-11e9-9590-0a580a2c0406", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Modified", GitTag:"1.2.0-rc.2-12-gf829550"}
```

```
▶ k version
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.9", GitCommit:"2e808b7cb054ee242b68e62455323aa783991f03", GitTreeState:"clean", BuildDate:"2020-01-18T23:33:14Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"15+", GitVersion:"v1.15.9-gke.24", GitCommit:"39e41a8d6b7221b901a95d3af358dea6994b4a40", GitTreeState:"clean", BuildDate:"2020-02-29T01:24:35Z", GoVersion:"go1.12.12b4", Compiler:"gc", Platform:"linux/amd64"}
```

```
▶ helm version --tls
Client: &version.Version{SemVer:"v2.9.1", GitCommit:"20adb27c7c5868466912eebdf6664e7390ebe710", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}
```

**How was Istio installed?**
```
        helm upgrade istio --install --tls install/kubernetes/helm/istio --namespace istio-system \
        --set tracing.enabled=true \
        --set tracing.provider=zipkin \
        --set pilot.traceSampling=1 \
        --set pilot.autoscaleMin=2 \
        --set pilot.replicaCount=2 \
        --set grafana.enabled=true \
        --set prometheus.retention=7d \
        --set kiali.enabled=true \
        --set kiali.dashboard.grafanaURL=http://grafana:3000 \
        --set kiali.dashboard.jaegerURL=http://jaeger-query:16686 \
        --set sidecarInjectorWebhook.rewriteAppHTTPProbe=true \
        --set global.mtls.enabled=true \
        --set global.outboundTrafficPolicy.mode=ALLOW_ANY \
        --set global.proxy.accessLogFile="/dev/stdout" \
        --set galley.enabled=true \
        --set gateways.enabled=true \
        --set gateways.istio-ingressgateway.enabled=true \
        --set mixer.policy.autoscaleMin=1 \
        --set mixer.policy.replicaCount=1 \
        --set mixer.telemetry.autoscaleMin=1 \
        --set mixer.telemetry.replicaCount=1 \
```

**Environment where bug was observed (cloud vendor, OS, etc)**
GKE

**Comment by howardjohn** [comment]
Can you try upgrade to a recent version of Istio and see if its resolved? Istio 1.2 is no longer supported and this bug may be fixed since then. See https://istio.io/news/support/announcing-1.2-eol/.

Thanks for the thorough bug report

**Comment by JimmyCYJ** [comment]
@pantelis-karamolegkos any updates? Could you try https://github.com/istio/istio/issues/23422#issuecomment-622174116 and see if it works now?
